SimTracker/TrackTriggerAssociation/interface/TTStubAssociationMap.h

```diff
@@ -41,48 +54,55 @@ class TTStubAssociationMap {
+
+  /// Set cluster <-> truth association object.
+  void setTTClusterAssociationMap(edm::RefProd<TTClusterAssociationMap<T>> aCluAssoMap) {
     theClusterAssociationMap = aCluAssoMap;
   }
 
-  /// Operations
-  edm::Ptr<TrackingParticle> findTrackingParticlePtr(edm::Ref<edmNew::DetSetVector<TTStub<T> >, TTStub<T> > aStub) const;
-  std::vector<edm::Ref<edmNew::DetSetVector<TTStub<T> >, TTStub<T> > > findTTStubRefs(
-      edm::Ptr<TrackingParticle> aTrackingParticle) const;
-
-  /// MC Truth methods
-  bool isGenuine(edm::Ref<edmNew::DetSetVector<TTStub<T> >, TTStub<T> > aStub) const;
-  bool isCombinatoric(edm::Ref<edmNew::DetSetVector<TTStub<T> >, TTStub<T> > aStub) const;
-  bool isUnknown(edm::Ref<edmNew::DetSetVector<TTStub<T> >, TTStub<T> > aStub) const;
+  /// Get principle TP associated to a stub. (Non-NULL if isGenuine() below is true).
+  /// (N.B. There is no function returning all TP associated to a stub).
+  /// (P.S. As this function only returns principle TP, it is not used when constructing
+  ///  the TTTrackAssociationMap).
+  TrackingParticlePtr findTrackingParticlePtr(TTStubRefT<T> aStub) const;
+
+  /// Get all stubs associated to a TP.
+  /// (Even if the TP just contributes to one cluster in stub,
+  /// and even if their are other such TP, it is still listed here).
+  std::vector<TTStubRefT<T>> findTTStubRefs(TrackingParticlePtr aTrackingParticle) const;
+
+  ///--- Get quality of stub based on truth info.
+  /// (N.B. Both genuine & combinatoric stubs contribute to "genuine" L1 tracks
+  ///  associated by TTTrackAssociationMap).
+  /// (exactly 1 of following 3 functions is always true)
+
+  /// If both clusters are unknown, the stub is "unknown".
+  /// If only one cluster is unknown, the stub is combinatoric.
+  /// If both clusters are genuine, and are associated to the same (main) TrackingParticle,
+  /// the stub is "genuine".
+  /// If both clusters are genuine, but are associated to different (main) TrackingParticles,
+  /// the stub is "combinatoric".
+  /// If one cluster is combinatoric and the other is genuine/combinatoric, and they both share exactly
+  /// one TrackingParticle in common, then the stub is "genuine". (The clusters can have other
+  /// TrackingParticles besides the shared one, as long as these are not shared). If instead the clusters
+  /// share 0 or ≥2 TrackingParticles in common, then the stub is "combinatoric".
+
+  bool isGenuine(TTStubRefT<T> aStub) const;
+  bool isCombinatoric(TTStubRefT<T> aStub) const;
+  bool isUnknown(TTStubRefT<T> aStub) const;
 
 private:
   /// Data members
-  std::map<edm::Ref<edmNew::DetSetVector<TTStub<T> >, TTStub<T> >, edm::Ptr<TrackingParticle> >
-      stubToTrackingParticleMap;
-  std::map<edm::Ptr<TrackingParticle>, std::vector<edm::Ref<edmNew::DetSetVector<TTStub<T> >, TTStub<T> > > >
-      trackingParticleToStubVectorMap;
-  edm::RefProd<TTClusterAssociationMap<T> > theClusterAssociationMap;
+  MapStubToTP<T> stubToTrackingParticleMap;
```

Comment:
member variable names should end in underscore

Reply:
done